lib/Agrammon/OutputFormatter/GUI.pm6

```diff
@@ -18,47 +20,34 @@ sub output-for-gui(Agrammon::Model $model,
-                my $formattedValue = ($format && $value.defined) ?? sprintf($format, $value)
-                                             !! $value;
-                push @records, %(
-                    format    => $format,
-                    print     => $model.output-print($module, $output),
-                    order     => $model.output-order($module, $output),
-                    labels    => $model.output-labels($module, $output),
-                    units     => $model.output-units($module, $output),
-                    fullValue => $value,
-                    value     => $formattedValue,
-                    var       =>  $module ~ '::' ~ $output,
-                );
+                my $var = $module ~ '::' ~ $output;
+                push @records, make-record($module, $output, $model, $raw-value, $var);
+                if $include-filters {
+                    my $value = flat-value($raw-value);
+                    if $raw-value ~~ Agrammon::Outputs::FilterGroupCollection && $raw-value.has-filters {
+                        push-filters(@records, $module, $output, $model, $raw-value, $var);
+                    }
+                }
             }
         }
         when Array {
             for sorted-kv($_) -> $instance-id, %instance-outputs {
                 for sorted-kv(%instance-outputs) -> $fq-name, %values {
                     my $q-name = $module ~ '[' ~ $instance-id ~ ']' ~ $fq-name.substr($module.chars);
                     for sorted-kv(%values) -> $output, $raw-value {
-                        my $format = $model.output-format($fq-name, $output);
-                        my $value = flat-value($raw-value);
-                        my $formattedValue = ($format  && $value.defined) ?? sprintf($format, $value)
-                                                     !! $value;
-                        push @records, %(
-                            format    => $format,
-                            print     => $model.output-print($fq-name, $output),
-                            order     => $model.output-order($fq-name, $output),
-                            labels    => $model.output-labels($fq-name, $output),
-                            units     => $model.output-units($fq-name, $output),
-                            fullValue => $value,
-                            value     => $formattedValue,
-                            var       =>  $q-name ~ '::' ~ $output,
-                        );
-
+                        my $var = $q-name ~ '::' ~ $output;
+                        push @records, make-record($fq-name, $output, $model, $raw-value, $var);
+                        if $include-filters {
+                            my $value = flat-value($raw-value);
+                            if $raw-value ~~ Agrammon::Outputs::FilterGroupCollection && $raw-value.has-filters {
+                                push-filters(@records, $fq-name, $output, $model, $raw-value, $var);
```

Comment:
I also wonder a bit if the filtered values should be part of the record of the value that they are breaking down, rather than just pushed as extra, flat, records. Maybe it's a bigger refactor than you want to do in the frontend right now, but it would make it easy to do stuff like making the breakdown by filters expandable/collapsable, since you'd know which filter values were the breakdown of which total values quite easily.

Reply:
I think I'd prefer to do the frontend refactoring later. There is another issue that might be worthwhile doing. Right now I am sending all the outputs to the frontend and the filter them there for the various "reports".
It might be more sensible to only send the outputs on demand, when they are really needed for a report the user actually looks at.
Not really all that relevant as there is not all that much data, so not decided yet what makes most sense.